### PR TITLE
feat!: Upgrade `Kraftfile` specification to v0.6

### PR DESCRIPTION
**Pull request by nderjung**
### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

This PR introduces a new `Kraftfile` specification which makes new adjustments to the user's workflow such that it is easier to run applications in binary compatibility mode and with a template.  At a high-level, the following adjustments are made to the specification:

- A new top-level `runtime` element to be used for defining the user of an elfloader or existing pre-built unikernel image.  Similar to other component-type elements like `unikraft` and `libraries`, the `runtime` element is intended to allow for a short-hand and long-hand syntax where a `source`, `version` and set of `kconfig` options can be used to qualify the elfloader application.
- Abbreviations of the `specification`, `architecture` and `platform` element names such that can be used in short-hand syntax as `spec`, `arch` and `plat`, respectively.
- Top-level `volumes` element which is used for defining user-specified paths which are to be used during runtime of a unikernel.
- Top-level `rootfs` element which is used for defining the root filesystem of the user.  The provided implementation allows a user to serialize a filesystem from three new contexts: an existing CPIO archive (passthrough), a path to a directory, and a `Dockerfile` which is constructed via [BuildKit](https://github.com/moby/buildkit).
- Top-level `cmd` which is a string or list which represents the command-line argument to execute for the unikernel as a way to porclain the Kraftfile and the intended application usecase.

Aside from these introductions to the v0.6 specification, the underlying parser is adjusted to handle reading both the specifications and defaulting to v0.6 as the latest.  These adjustments are subsequently implemented through the PR.

#### A. Introduction of short-hand syntaxes

1. In addition to bumping the specification version to v0.6, the `specification` element can be abbreviated to `spec`:

   ```yaml
   spec: v0.6
   ```

2. Within the list of `targets`, the `architecture` and `platform` attributes can be abbreviated to `arch` and `plat`, respectively, and be used interchangeably:

   ```yaml
   targets:
   - plat: qemu
     arch: x86_64
   ```

   ```yaml
   targets:
   - plat: qemu
     architecture: x86_64
   ```

   ```yaml
   targets:
   - platform: qemu
     arch: x86_64
   ```

3. The list of targets can now accept an even shorter syntax where only the architecture and platform are desired in the list:

   ```yaml
   targets:
   - qemu/x86_64
   ```

   This shorthand syntax can be mixed with full target elements:

   ```yaml
   targets:
   - qemu/x86_64
   - platform: qemu
     architecture: arm64
   - plat: fc
     arch: x86_64
   ```

#### C. Introduction of a `cmd` element for specifying program operation

The `cmd` is array or string element which can be used for setting default arguments to be used during the instantiation of a new unikernel instance:

1. Specified as an in-line array:
   ```yaml
   spec: v0.6

   cmd: ["-c", "/nginx/conf/nginx.conf"]
   ```

2. Specified as a multi-line array:
   ```yaml
   spec: v0.6

   cmd:
   - -c
   - /nginx/conf/nginx.conf
   ```

3. Specified as a string:
   ```yaml
   spec: v0.6

   cmd: "-c /nginx/conf/nginx.conf"
   ```

#### C. Introduction to `runtime` element for accessing pre-built images

The `runtime` element is a new top-level element which has been designed to ease the declaration of a previously constructed unikernel image in the project.  In many cases, specifically instances where the user is writing in a high-level language such as Python or where they wish to use a pre-built Linux userpace image, the user does not care for wanting to build a unikernel image from scratch.  Instead, the goal is to leverage pre-built unikernel images which are distributed using the [OCI Package Manager](https://github.com/unikraft/kraftkit/pull/339). 

Similar to the top-level `unikraft` and `template` elements (see below), the `runtime` element can be specified using a short-hand syntax or a long-hand syntax for greater control or preference.

1. In short-hand syntax, simply referencing the pre-built unikernel image is possible by declaring:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   ```
   In the above example, the `runtime` element refers to an OCI archive for NGINX.  In a `kraft run` scenario, this is simply pulled and executed verbatim.  Paired with the ability to customize the `rootfs` element or `volumes` elemen, the `runtime` element becomes more pragmatic:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   
   rootfs: ./
   ```
   Or with a volume, which simply "overwrites" an existing directory already located within NGINX's root filesystem that is distributed as part of the OCI image:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/nginx:latest
   
   volumes:
   - ./html:/nginx/html
   ```
   In the above snippet, the user customizes only the directory which serves HTML with a path `./html` which is located on their host.
   
   Note that this usecase is not limited to a single application and applies to the ELF Loader application and customizing the loaded application which is part of a root filesystem:
   ```yaml
   spec: v0.6
   
   # Default ELF Loader application
   runtime: loaders.unikraft.org/default:latest 
   
   # Path must contain the Linux ELF binary, e.g. at ./rootfs/helloworld
   rootfs: ./rootfs
   
   cmd: ["/helloworld"]
   ```
   or a high-level programming language such as python3:
   ```yaml
   spec: v0.6
   
   runtime: unikraft.org/python3:latest
   
   rootfs: ./Dockerfile
   
   cmd: ["/main.py"]
   ```
   In the above snippet, the user has created a `Dockerfile` which contains the necessary root filesystem and included their application at `/main.py`.

2. The long-hand syntax allows specifically setting `source`, `version` and `kconfig` options which are requirements of the runtime:
   ```yaml
   spec: v0.6
   
   runtime:
     source: unikraft.org/nginx
     version: latest
     kconfig:
       CONFIG_LIB9PFS: 'y'
    ```
    In the above example, the a pre-built NGINX unikernel has been requested as the runtime with the explicit option of having the 9P Filesystem enabled.

#### D. Fully realizing the `template` element

There were several issues related to the `template` element, including the fact that it was not properly seeded, or transformed, from the `Kraftfile` into the application structure. This has now been tested end-to-end and enables:

1. Workflows where a user may wish to reference an existing application repository and make customizations, for example, with the [ELF Loader](https://github.com/unikraft/app-elfloader.git) application, making changes to the core configuration and adding additional targets:
   ```yaml
   spec: v0.6
   
   template: https://github.com/unikraft/app-elfloader.git
   
   unikraft:
     version: stable
     kconfig:
       CONFIG_LIBUKLIBPARAM: 'y'
       CONFIG_LIBNOLIBC: 'y'
       CONFIG_LIBNOLIBC_UKDEBUG_ASSERT: 'y'
       CONFIG_LIBPOSIX_EVENT: 'y'
       CONFIG_LIBPOSIX_FUTEX: 'y'
       CONFIG_LIBPOSIX_MMAP: 'y'
       CONFIG_LIBPOSIX_PROCESS: 'y'
       CONFIG_LIBPOSIX_PROCESS_PIDS: 'y'
       CONFIG_LIBPOSIX_PROCESS_MAX_PID: 31
       CONFIG_LIBPOSIX_PROCESS_CLONE: 'y'
       CONFIG_LIBPOSIX_SOCKET: 'y'
       CONFIG_LIBPOSIX_SYSINFO: 'y'
       CONFIG_LIBPOSIX_TIME: 'y'
       CONFIG_LIBPOSIX_USER: 'y'
       CONFIG_LIBUKSIGNAL: 'y'
       CONFIG_LIBSYSCALL_SHIM: 'y'
       CONFIG_LIBVFSCORE_AUTOMOUNT_ROOTFS: 'y'
       CONFIG_LIBVFSCORE_ROOTFS_INITRD: 'y'
       CONFIG_LIBVFSCORE_ROOTFS: "initrd"
       CONFIG_APPELFLOADER_VFSEXEC_EXECBIT: 'n'

   targets:
   - plat: qemu
     arch: x86_64
     kconfig:
       CONFIG_VIRTIO_BUS: 'n'
   - plat: firecracker
     arch: x86_64
   ```

3. Similar to other component elements (e.g. `unikraft` and `libraries`), the `template` element can be expressed in a longer format:
   ```yaml
   spec: v0.6
   
   template:
     source: https://github.com/unikraft/app-elfloader.git
     version: stable
   ```

### New workflow scenarios

The following workflow scenarios were used during testing.

#### A. Using a pre-built unikernel and customizing the root filesystem

Customizing the whole root filesystem is now possible by supplying either:

1. A path representing a prebuilt CPIO archive file:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./initramfs.cpio
   ```

4. A directory which will be dynamically serialized into a CPIO archive:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./rootfs/
   ```

5. Or, a `Dockerfile` which will be built via BuildKit and then serialized into a new CPIO archive:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   rootfs: ./Dockerfile
   ```
   To use BuildKit, set the address within the new `buildkit_host` user setting, the new environmental variable `KRAFTKIT_BUILDKIT_HOST` or the command-line flag `--buildkit-host`.  This defaults to `unix:///run/buildkit/buildkitd.sock`.

#### B. Using a pre-built unikernel and mounting a path within the root filesystem

1. Specifying via short-hand syntax a map between a directory on the host and a path within the instance:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   volumes:
   - ./src:/nginx/html
   ```

2. Using a more verbose (long-hand) syntax and mapping between the mount on the host and a path within the instance:
   ```yaml
   spec: v0.6

   runtime: unikraft.org/nginx:latest

   volumes:
   - source: ./src
     destination: /nginx/html
   ```

### CODE REVIEW

File: unikraft/app/application.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ type Application interface {
 	Template() template.Template
 
 	// Libraries returns the application libraries' configurations
-	Libraries(ctx context.Context) (lib.Libraries, error)
+	Libraries(ctx context.Context) (map[string]*lib.LibraryConfig, error)
```
Comment (craciunoiuc) [inline unikraft/app/application.go:56]: In the commit (commit 1) message for this change: `ambigiuity` -> `ambiguity`

File: oci/image.go [hunk 2]
```diff
@@ -205,6 +205,11 @@ func (image *Image) SetOSFeature(_ context.Context, feature ...string) {
 	image.config.OSFeatures = append(image.config.OSFeatures, feature...)
 }
 
+// Set the entrypoint of the image.
+func (image *Image) SetEntrypoint(_ context.Context, entrypoint string) {
```
Comment (craciunoiuc) [inline oci/image.go]: In commit (`feat: Populate application entrypoint from Kraftfile`) message `packaginf` -> `packaging`

File: action.yml [hunk 3]
```diff
@@ -55,8 +55,8 @@ inputs:
   args:
     description: Arguments to pass to the unikernel.
     required: false
-  initrd:
-    description: Include an initrd at path.
+  rootfs:
```
Comment (craciunoiuc) [inline action.yml:58]: In commit `feat!: Introduce dynamic initramfs serialization for root filesystems` `underelying` -> `underlying`

File: initrd/dockerfile.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+)
+
+type dockerfile struct {
+	opts       InitrdOptions
+	dockerfile string
+	files      []string
+	workdir    string
+}
+
+// NewFromDockerfile accepts an input path which represents a Dockerfile that is
```
Comment (craciunoiuc) [inline initrd/dockerfile.go]: `is` -> ` `

**Review by craciunoiuc — CHANGES_REQUESTED** [review]
Went through the whole PR

Feels like it needs to be split up into at least 2 PRs, maybe 3

1 with the fixes, 1 with the spec update and 1 with the initrd work

File: initrd/dockerfile.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+			return fmt.Errorf("could not read file: %w", err)
+		}
+
+		if err := writer.WriteHeader(&cpio.Header{
+			Name:     internal,
+			Linkname: targetLink,
+			Mode:     cpio.FileMode(d.Type().Perm()),
+			Size:     info.Size(),
+		}); err != nil {
+			return fmt.Errorf("could not writer CPIO header: %w", err)
```
Comment (craciunoiuc) [inline initrd/dockerfile.go]: `write`

File: initrd/dockerfile.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+		}
+		if err != nil {
+			return fmt.Errorf("could not read file: %w", err)
+		}
+
+		if err := writer.WriteHeader(&cpio.Header{
+			Name:     internal,
+			Linkname: targetLink,
+			Mode:     cpio.FileMode(d.Type().Perm()),
+			Size:     info.Size(),
```
Comment (craciunoiuc) [inline initrd/dockerfile.go:204]: Are permissions in initramfs fixed now?

Comment (nderjung) [reply initrd/dockerfile.go:204]: Maybe 🤷

File: unikraft/target/transform.go [hunk 7]
```diff
@@ -17,15 +18,31 @@ import (
...
+		t.name = uk.UK_NAME
+	}
+
 	switch value := data.(type) {
-	case map[string]interface{}:
-		uk := unikraft.FromContext(ctx)
-		t := TargetConfig{}
+	case string:
+		split := strings.SplitN(value, "/", 2)
+
```
Comment (craciunoiuc) [inline unikraft/target/transform.go:30]: Can cause sigsegv it's not tested

File: oci/pack.go [hunk 8]
```diff
@@ -171,7 +171,7 @@ func NewPackageFromTarget(ctx context.Context, targ target.Target, opts ...packm
 			WithLayerAnnotation(AnnotationKernelInitrdPath, WellKnownInitrdPath),
 		)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("could build layer from file: %w", err)
```
Comment (craciunoiuc) [inline oci/pack.go:174]: `not`

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Pull request merged**